torch_xla/csrc/data_ops.cpp

```diff
@@ -2,24 +2,52 @@
 #include "helpers.h"
 #include "tensorflow/compiler/xla/xla_client/debug_macros.h"
 
+#include <functional>
+#include <numeric>
+
 namespace torch {
 namespace jit {
 
 namespace {
 
-// Graph nodes specify -1 for unknown dimensions. Return true iff all dimension
-// sizes are positive.
-bool IsCompleteShape(const std::vector<int64_t>& dim_sizes) {
-  return std::all_of(dim_sizes.begin(), dim_sizes.end(),
-                     [](const int64_t dim_size) { return dim_size >= 0; });
+// For input_sizes and a potentially incomplete output_sizes, return a complete
+// output shape. The complete output shape has same total number of elements as
+// input_sizes and matches output_sizes in all dimensions except for at most
+// one, which can be inferred and stored as -1 in output_sizes.
+std::vector<int64_t> GetCompleteShape(
+    const std::vector<int64_t>& output_sizes,
+    const std::vector<xla::int64>& input_sizes) {
+  c10::optional<size_t> incomplete_dim;
+  int64_t incomplete_element_count = 1;
+  for (size_t dim = 0; dim < output_sizes.size(); ++dim) {
+    const auto dim_size = output_sizes[dim];
+    if (dim_size < 0) {
+      XLA_CHECK(!incomplete_dim) << "More than one incomplete dimension found";
```

Comment:
It helps to pipe stuff into the CHECK. We might get more useful debugging info.
Example:

XLA_CHECK(!incomplete_dim) << "More than one incomplete dimension found: " << *incomplete_dim << " and " << dim;